packages/xod-project/src/flatten.js

```diff
@@ -626,6 +626,143 @@ const removeTerminalsAndPassPins = ([nodes, links]) => {
   return [newNodes, newLinks];
 };
 
+const isLinkOutputNodeIdOneOf = R.curry((terminalNodeIds, link) =>
+  R.compose(R.contains(R.__, terminalNodeIds), Link.getLinkOutputNodeId)(link)
+);
+
+// It collects the full chain of nodeIds and links from the lowermost to
+// uppermost links.
+// Links collected in order from lowermost to uppermost.
+// NodeIds collected to omit Nodes that replaced with a new link.
+// If the last link in array linked with terminal node — it calls recursively.
+// :: [NodeId] -> [Link] -> [[NodeId], [Link]]
+const traverseUpAndCollectTerminalChain = R.curry(
+  (terminalNodeIds, links, collected) => {
+    const curLink = R.compose(R.last, R.last)(collected);
+
+    if (isLinkOutputNodeIdOneOf(terminalNodeIds, curLink)) {
+      // collect nodeId + next link & run recursion
+      const nextNodeId = Link.getLinkOutputNodeId(curLink);
+      const nextLink = R.find(
+        R.compose(R.equals(nextNodeId), Link.getLinkInputNodeId),
+        links
+      );
+      return R.compose(
+        traverseUpAndCollectTerminalChain(terminalNodeIds, links),
+        R.over(R.lensIndex(0), R.append(nextNodeId)),
+        R.over(R.lensIndex(1), R.append(nextLink))
+      )(collected);
+    }
+    return collected;
+  }
+);
+
+// Replaces generic internal terminals (`xod/internal/terminal-t1`) with
+// cast node if needed or just create a direct link from uppermost output pin
+// to the lowermost input pin, which are connected through any amount of internal
+// generic terminals (any nesting level).
+// :: Project -> [[PatchPath, Patch]] -> [[Node], [Link]] -> [[Node], [Link]]
+const replaceGenericTerminalsWithCastNodes = R.curry(
+  (project, patchTuples, [nodes, links]) => {
+    const terminalNodeIds = R.compose(
+      R.map(Node.getNodeId),
+      R.filter(
+        R.both(
+          isInternalTerminalNode,
+          R.compose(R.test(/t1$/), Node.getNodeType)
```

Comment:
```suggestion
          R.compose(isGenericType, Node.getNodeType)
```

(`isGenericType` is defined in `./utils.js`)

Reply:
Oh. No. It works only with common terminals, not internals 😬 
I extracted it into a function with a clear name `isGenericInternalTerminal` and replaced this `R.test` with a new function in PatchPathUtils.js `isInternalGenericPatchPath :: PatchPath -> Boolean`.